### ISSUE TITLE
Mismatch in LLaMAMoE litgpt and hf implementation for Mixtral

### ISSUE THREAD

**Issue by ysjprojects** [Lightning-AI/litgpt#2013]
### Bug description

litgpt:
```
probs, indices = torch.topk(router, self.config.n_expert_per_token)  # (B*T, n_expert_per_token)
probs = probs.softmax(dim=1, dtype=torch.float).to(dtype=x.dtype)
```

hf:
```
routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
```

original hf implementation applies softmax before selecting the topk while litgpt implementation does it after. Since softmax is monotonic the ordering and therefore the topk experts are preserved. However, the probability values will be different under the same condition.

### What operating system are you using?

macOS

### LitGPT Version

```
Version: 0.5.4.dev1
```

**Comment by t-vi** [comment]
In the PR, we also have a renormalization
```python
routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
routing_weights = routing_weights / routing_weights.sum(dim=1, keepdim=True)
```

but then we have a computation that is mathematically equivalent to the much more efficient 

```
routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
```

The softmax in both cases will give the same vector up to a normalization factor (the first so that the full vector sums to one, the second such that the selected vector sums to one), but then the division by the sum will give the same result.

To my mind, our version is a legitimate optimization.
